docs/design/alto/client-interaction.md

```diff
@@ -0,0 +1,136 @@
+their counterparts. On top of that, they contain the following two extra
+response parameters:
+
+- `tpcPorts`: List of the exposed TPC ports, or null if the TPC is disabled
+on the server-side.
+- `tpcToken`: The token to be used during the authentication of the TPC 
+connections, or null if the TPC is disabled on the server-side.
+
+The client uses the classic Hazelcast port to authenticate, as usual. The
+same discovery mechanisms are also used to specify the addresses to connect,
+and there are no TPC specific changes in that area.
+
+When the TPC-enabled client connects to the classic Hazelcast port of the TPC-enabled
+cluster with these new experimental codecs, on top of the regular authentication
+procedure and response, these two extra fields are returned to the client. Only
+then the client knows about the exposed TPC ports of the cluster member.
+
+The ports are ordered list with the following sample format:
+
+```
+[11000, 11001, 11002, ...]
+```
+
+The client will use the hostname of the classic Hazelcast connection and these
+ports to establish connections to each of the event loops.
+
+The returned token is a 64-bytes long byte array, generated using 
+[Secure Random](https://docs.oracle.com/javase/8/docs/api/java/security/SecureRandom.html).
+A new token is generated for each authenticated client connection to any of the
+cluster members. So, there is a separate token associated with each successful
+client connection to the classic Hazelcast ports of the cluster members.
+
+The connection to the TPC ports follows a similar route as the connections to
+classic ports.
+
+First, 3-bytes long protocol identifier is sent. (bytearray representation of `CP2`).
+
+Then, the client authenticates to the TPC event loop using the new 
+`ExperimentalTpcAuthenticationCodec`, which has the following definition.
+
+```yaml
+- id: 3
+  name: tpcAuthentication
+  since: 2.6
+  doc: |
+    Makes an authentication request to TPC channels.
+  request:
+    retryable: true
+    partitionIdentifier: -1
+    params:
+      - name: uuid
+        type: UUID
+        nullable: false
+        since: 2.6
+        doc: |
+          UUID of the client.
+      - name: token
+        type: byteArray
+        nullable: false
+        since: 2.6
+        doc: |
+          Authentication token bytes for the TPC channels
+  response: {}
+```
+
+In that authentication request, the client sends the necessary information
+to identify itself (its UUID), and the TPC token associated with the
+client connection to the classic Hazelcast port, to prove its identity.
+
+If the token associated with the client UUID matches the token sent, the 
+server returns a successful authentication response to the client (an empty message). 
+
+The server has the following protections for unauthorized connection attempts:
+
+- If anything other than `CP2` is sent as the protocol bytes initially, the 
+connection is closed immediately.
```

Comment:
I haven't but it might make sense. Will it be OK to discuss that in the 5.4 cycle, as I don't want to change too many things in this PR and we are a bit short in time?